hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/holder/PagingPredicateHolder.java

```diff
@@ -116,7 +116,7 @@ private static <K, V> PagingPredicateHolder ofInternal(@Nonnull PartitionPredica
                                                            SerializationService serializationService) {
         PagingPredicateImpl<K, V> pagingPredicate = (PagingPredicateImpl<K, V>) partitionPredicate.getTarget();
 
-        Data partitionKeyData = serializationService.toData(partitionPredicate.getPartitionKey());
+        Data partitionKeyData = serializationService.toData(partitionPredicate.getPartitionKeys());
```

Comment:
I feel like to make it work, we need to add a new parameter to the https://github.com/hazelcast/hazelcast-client-protocol/blob/master/protocol-definitions/custom/Custom.yaml#L852 with the type of `List_Data`, and set both the `partitionKeyData` and the new parameter so that both the old and new members can continue working. Note that, the `since` value of the parameter should be `2.5` for the 5.2 release

Reply:
A similar thing applies to the read path (where we read client messages from the old clients like 5.0)

Reply:
That's exactly what has caused the two tests to fail, and why I was originally so cautious about touching the client code (it would suffice to have server side support only for our usecase).

Do you think it's worth doing the client side code in a separate PR @mdumandag?

Reply:
I would vote for doing it here. WDYT @ahmetmircik?

Reply:
we can do it here, no need a separate PR.

Reply:
I've managed to get the codecs working with backwards compatible request/response.

The bug causing these two unit test failures isn't because the return type is incompatible (MapKeySet already returns List_Data because querying a single partition can still result in multiple return values).

The bug was introduced when I changed `AbstractMapQueryMessageTask.invokeOnPartitions` to use the map service query engine for a keyset:

```QueryResult result = mapServiceContext.getQueryEngine(name).execute(query, Target.createPartitionTarget(query.getPartitionIdSet()));```

When this code is executed it adjusts the predicate specified on the query in `QueryEngineImpl.adjustQuery` 
```private Query adjustQuery(Query query) {
        IterationType retrievalIterationType = getRetrievalIterationType(query.getPredicate(), query.getIterationType());
        Query adjustedQuery = Query.of(query).iterationType(retrievalIterationType).build();
        if (adjustedQuery.getPredicate() instanceof PagingPredicateImpl) {
            ((PagingPredicateImpl) adjustedQuery.getPredicate()).setIterationType(query.getIterationType());
        } else {
            if (adjustedQuery.getPredicate() == Predicates.alwaysTrue()) {
                queryResultSizeLimiter.precheckMaxResultLimitOnLocalPartitions(adjustedQuery.getMapName());
            }
        }
        return adjustedQuery;
    }```

This adjustment changes `iterationType` on `PagingPredicateImpl` instance to match the iteration type specified by the query object (in this case it changes KEY to ENTRY). 

Looking at the code `MapKeySetWithPagingPredicateTask.getIterationType` is implemented as: 
```@Override
    protected IterationType getIterationType() {
        return IterationType.ENTRY;
    }``` 

This means there has always been an inconsistency between the `PagingPredicateImpl` and `MapKeySetWithPagingPredicateTask`, because the old code never executed a map query via the service it was never noticed.

I'm proposing we change ```MapKeySetWithPagingPredicateTask.getIterationType``` to return `IterationType.KEY`